# TSE Data Analysis

In [1]:
library(dplyr)
library(ggplot2)


Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union




In [2]:
prs <- read.csv('../data/prs.csv', stringsAsFactors=TRUE)
head(prs)
summary(prs)

cursor       owner  name  repoUrl                         clients
1 Y3Vyc29yOjEx google guava https://github.com/google/guava 304728 
2 Y3Vyc29yOjEx google guava https://github.com/google/guava 304728 
3 Y3Vyc29yOjEx google guava https://github.com/google/guava 304728 
4 Y3Vyc29yOjEx google guava https://github.com/google/guava 304728 
5 Y3Vyc29yOjEz apache dubbo https://github.com/apache/dubbo  50218 
6 Y3Vyc29yOjEz apache dubbo https://github.com/apache/dubbo  50218 
  prUrl                                     
1 https://github.com/google/guava/pull/6232 
2 https://github.com/google/guava/pull/6225 
3 https://github.com/google/guava/pull/6218 
4 https://github.com/google/guava/pull/6188 
5 https://github.com/apache/dubbo/pull/10929
6 https://github.com/apache/dubbo/pull/10926
  title                                                                      
1 Fixes for issues #3965, #1710, #5882 and #5773                             
2 added testput_nullvaluesupported to test                                   
3 PR: Fix #3720                                                              
4 Fix Spliterator implementations for sorted collections                     
5 InstantiationAwareBeanPostProcessorAdapter即将废弃，不再兼容Spring Boot 3.x
6 Add some logs                                                              
  number state draft ⋯ baseRepo     baseRef baseRefPrefix headRepo        
1  6232  OPEN  false ⋯ google/guava master  refs/heads/   u6873285/guava  
2  6225  OPEN  false ⋯ google/guava master  refs/heads/   alvonellos/guava
3  6218  OPEN  false ⋯ google/guava master  refs/heads/   yonghao97/guava 
4  6188  OPEN  false ⋯ google/guava master  refs/heads/   kilink/guava    
5 10929  OPEN  false ⋯ apache/dubbo 3.1     refs/heads/   chang6666/dubbo 
6 10926  OPEN  false ⋯ apache/dubbo 3.2     refs/heads/   AlbumenJ/dubbo  
  headRef                                                headRefPrefix
1 master                                                 refs/heads/  
2 #6217_Add_test_that_putIfAbsent_replaces_a_null_value_ refs/heads/  
3 fix_3720                                               refs/heads/  
4 sorted-collection-spliterator-fix                      refs/heads/  
5 3.1                                                    refs/heads/  
6 1112_add_log                                           refs/heads/  
  createdAt  publishedAt mergedAt closedAt
1 2022-10-29 2022-10-29  NA       NA      
2 2022-10-23 2022-10-23  NA       NA      
3 2022-10-17 2022-10-17  NA       NA      
4 2022-09-28 2022-09-28  NA       NA      
5 2022-11-13 2022-11-13  NA       NA      
6 2022-11-12 2022-11-12  NA       NA

              cursor              owner           name     
 Y3Vyc29yOjIxNA==: 343   apache      :1066   trino  : 343  
 Y3Vyc29yOjI0MA==: 214   trinodb     : 343   alluxio: 214  
 Y3Vyc29yOjM4OA==: 144   Alluxio     : 214   hudi   : 144  
 Y3Vyc29yOjU1    : 139   OpenAPITools:  78   flink  : 139  
 Y3Vyc29yOjExMg==: 125   alibaba     :  56   pulsar : 125  
 Y3Vyc29yOjk4    :  93   keycloak    :  56   hadoop :  93  
 (Other)         :1470   (Other)     : 715   (Other):1470  
                               repoUrl        clients       
 https://github.com/trinodb/trino  : 343   Min.   :    132  
 https://github.com/Alluxio/alluxio: 214   1st Qu.:   1589  
 https://github.com/apache/hudi    : 144   Median :   3272  
 https://github.com/apache/flink   : 139   Mean   :  28792  
 https://github.com/apache/pulsar  : 125   3rd Qu.:  22544  
 https://github.com/apache/hadoop  :  93   Max.   :1374859  
 (Other)                           :1470                    
                                

## Popular active Java libraries on GitHub, as of 2022-11-13
  - Repositories written in Java
  - With a pom.xml file at the root
  - With >= 1000 stars
  - With >= 100 clients (all packages included)
  - Last push activity >= 2022-10-13 (1 month)
  - PR created >= 2022-10-13 (1 month)
  - PR merged >= 2022-09-13 (2 months)
  - !archived, !disabled, !fork, !mirror, !locked

In [3]:
libs <- distinct(prs[, c('owner', 'name', 'clients')])
head(libs)
summary(libs)

owner     name       clients
1 google    guava      304728 
2 apache    dubbo       50218 
3 jeecgboot jeecg-boot   1085 
4 alibaba   arthas        884 
5 netty     netty      120116 
6 alibaba   druid      309589

          owner           name        clients       
 apache      : 33   druid   :  2   Min.   :    132  
 alibaba     : 10   activemq:  1   1st Qu.:   1501  
 google      :  4   Activiti:  1   Median :   6420  
 spring-cloud:  4   alluxio :  1   Mean   :  56016  
 kiegroup    :  3   antlr4  :  1   3rd Qu.:  42081  
 sofastack   :  3   arthas  :  1   Max.   :1374859  
 (Other)     :100   (Other) :150                    

## Interesting pull requests
  - Opened, unmerged, and unlocked
  - Affects Java files
  - Updated >= 2022-10-13 (1 month)

In [4]:
head(prs)
summary(prs)

cursor       owner  name  repoUrl                         clients
1 Y3Vyc29yOjEx google guava https://github.com/google/guava 304728 
2 Y3Vyc29yOjEx google guava https://github.com/google/guava 304728 
3 Y3Vyc29yOjEx google guava https://github.com/google/guava 304728 
4 Y3Vyc29yOjEx google guava https://github.com/google/guava 304728 
5 Y3Vyc29yOjEz apache dubbo https://github.com/apache/dubbo  50218 
6 Y3Vyc29yOjEz apache dubbo https://github.com/apache/dubbo  50218 
  prUrl                                     
1 https://github.com/google/guava/pull/6232 
2 https://github.com/google/guava/pull/6225 
3 https://github.com/google/guava/pull/6218 
4 https://github.com/google/guava/pull/6188 
5 https://github.com/apache/dubbo/pull/10929
6 https://github.com/apache/dubbo/pull/10926
  title                                                                      
1 Fixes for issues #3965, #1710, #5882 and #5773                             
2 added testput_nullvaluesupported to test                                   
3 PR: Fix #3720                                                              
4 Fix Spliterator implementations for sorted collections                     
5 InstantiationAwareBeanPostProcessorAdapter即将废弃，不再兼容Spring Boot 3.x
6 Add some logs                                                              
  number state draft ⋯ baseRepo     baseRef baseRefPrefix headRepo        
1  6232  OPEN  false ⋯ google/guava master  refs/heads/   u6873285/guava  
2  6225  OPEN  false ⋯ google/guava master  refs/heads/   alvonellos/guava
3  6218  OPEN  false ⋯ google/guava master  refs/heads/   yonghao97/guava 
4  6188  OPEN  false ⋯ google/guava master  refs/heads/   kilink/guava    
5 10929  OPEN  false ⋯ apache/dubbo 3.1     refs/heads/   chang6666/dubbo 
6 10926  OPEN  false ⋯ apache/dubbo 3.2     refs/heads/   AlbumenJ/dubbo  
  headRef                                                headRefPrefix
1 master                                                 refs/heads/  
2 #6217_Add_test_that_putIfAbsent_replaces_a_null_value_ refs/heads/  
3 fix_3720                                               refs/heads/  
4 sorted-collection-spliterator-fix                      refs/heads/  
5 3.1                                                    refs/heads/  
6 1112_add_log                                           refs/heads/  
  createdAt  publishedAt mergedAt closedAt
1 2022-10-29 2022-10-29  NA       NA      
2 2022-10-23 2022-10-23  NA       NA      
3 2022-10-17 2022-10-17  NA       NA      
4 2022-09-28 2022-09-28  NA       NA      
5 2022-11-13 2022-11-13  NA       NA      
6 2022-11-12 2022-11-12  NA       NA

              cursor              owner           name     
 Y3Vyc29yOjIxNA==: 343   apache      :1066   trino  : 343  
 Y3Vyc29yOjI0MA==: 214   trinodb     : 343   alluxio: 214  
 Y3Vyc29yOjM4OA==: 144   Alluxio     : 214   hudi   : 144  
 Y3Vyc29yOjU1    : 139   OpenAPITools:  78   flink  : 139  
 Y3Vyc29yOjExMg==: 125   alibaba     :  56   pulsar : 125  
 Y3Vyc29yOjk4    :  93   keycloak    :  56   hadoop :  93  
 (Other)         :1470   (Other)     : 715   (Other):1470  
                               repoUrl        clients       
 https://github.com/trinodb/trino  : 343   Min.   :    132  
 https://github.com/Alluxio/alluxio: 214   1st Qu.:   1589  
 https://github.com/apache/hudi    : 144   Median :   3272  
 https://github.com/apache/flink   : 139   Mean   :  28792  
 https://github.com/apache/pulsar  : 125   3rd Qu.:  22544  
 https://github.com/apache/hadoop  :  93   Max.   :1374859  
 (Other)                           :1470                    
                                

## Results

In [5]:
results <- read.csv('../data/results.csv', stringsAsFactors=TRUE)
head(results)
summary(results)

base   breakingChanges brokenClients brokenUses changedFiles checkedClients
1 master 0               0             0          10           0             
2 master 0               0             0           1           0             
3 master 0               0             0           4           0             
4 master 0               0             0           2           0             
5 3.1    0               0             0           3           0             
6 3.1    0               0             0           1           0             
  errors                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
1 null[package] failed (1): [ERROR] COMPILATION ERROR : [ERROR] An unknown compilation problem occurred[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project guava: Compilation failure[ERROR] An unknown compilation problem occurred[ERROR] -> [Help 1][ERROR] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.[ERROR] Re-run Maven using the -X switch to enable full debug logging.[ERROR] [ERROR] For more information about the errors and possible solutions, please read the following articles:[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException[ERROR] [ERROR] After correcting the problems, you can resume the build with the command[ERROR]   mvn <args> -rf :guava[package] failed (1): [ERROR] COMPILATION ERROR : [ERROR] An unknown compilation problem occurred[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project guava: Compilation failure[ERROR] An unknown compilation problem occurred[ERROR] -> [Help 1][ERROR] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.[ERROR] Re-run Maven using the -X switch to enable full debug logging.[ERROR] [ERROR] For more information about the errors and possible solutions, please read the following articles:[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException[ERROR] [ERROR] After correcting the problems, you can resume the build with the command[ERROR]   mvn <args> -rf :guava
2 null[package] failed (1): [ERROR] COMPILATION ERROR : [ERROR] An unknown compilation problem occurred[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project guava: Compilation failure[ERROR] An unknown compilation probl

      base      breakingChanges     brokenClients       brokenUses     
 master :1884   Min.   :    0.000   Min.   : 0.0000   Min.   :   0.00  
 main   : 285   1st Qu.:    0.000   1st Qu.: 0.0000   1st Qu.:   0.00  
 dev    : 129   Median :    0.000   Median : 0.0000   Median :   0.00  
 develop:  79   Mean   :    9.588   Mean   : 0.1126   Mean   :   1.97  
 trunk  :  62   3rd Qu.:    0.000   3rd Qu.: 0.0000   3rd Qu.:   0.00  
        :  31   Max.   :17826.000   Max.   :91.0000   Max.   :1154.00  
 (Other): 204                                                          
  changedFiles    checkedClients   
 Min.   :   0.0   Min.   :  0.000  
 1st Qu.:   2.0   1st Qu.:  0.000  
 Median :   5.0   Median :  0.000  
 Mean   :  22.8   Mean   :  4.059  
 3rd Qu.:  13.0   3rd Qu.:  0.000  
 Max.   :3000.0   Max.   :337.000  
                                   
                                                                                                                                        

In [6]:
breaking <- subset(results, breakingChanges > 0)
nrow(breaking)
impacting <- subset(breaking, brokenUses > 0)
nrow(impacting)
nrow(results)

[1] 271

[1] 42

[1] 2674

Overall, 271 out of 2674 PRs are breaking (10.13%), and 42 have an impact on clients (1.57% of all, 15.5% of breaking ones).

In [7]:
all <- merge(prs, breaking, by=c('owner', 'name', 'number'))
write.csv(all, '../data/breaking-prs.csv')

## BreakBot reports
  - First, we fork the original repository in our own organization `breakbot-playground`
  - Then, we copy the PRs of interest in the fork
  - BreakBot runs automatically on these PRs to produce its reports
  - The `.github/breakbot.yml` configuration we use is as follows (100 clients per impacted package at most, with a minimum of 5 stars)
  
```
clients:
  top: 100
  stars: 5
```
